core/services/workflows/engine.go

```diff
@@ -148,73 +152,96 @@ func (e *Engine) triggerHandlerLoop(ctx context.Context) {
-	e.logger.Debugw("running consensus", "event", event)
-	cr := capabilities.CapabilityRequest{
-		Metadata: capabilities.RequestMetadata{
-			WorkflowID:          mockedWorkflowID,
-			WorkflowExecutionID: mockedExecutionID,
-		},
-		Inputs: &values.Map{
-			Underlying: map[string]values.Value{
-				// each node provides a single observation - outputs of mercury trigger
-				"observations": &values.List{
-					Underlying: []values.Value{event.Value},
+	ec := &executionState{
+		steps: map[string]*stepState{
+			trigger.Ref: {
+				outputs: &stepOutput{
+					value: event.Value,
 				},
 			},
 		},
-		Config: e.consensusConfig,
+		workflowID:  mockedWorkflowID,
+		executionID: mockedExecutionID,
 	}
-	chReports := make(chan capabilities.CapabilityResponse, 10)
-	newCtx, cancel := context.WithCancel(ctx)
-	defer cancel()
-	err := e.consensus.Execute(newCtx, chReports, cr)
+
+	consensus := e.workflow.Consensus[0]
+	err := e.handleStep(ctx, ec, consensus)
 	if err != nil {
-		return nil, err
+		e.logger.Errorf("error in handleConsensus %v", err)
+		return
 	}
-	select {
-	case resp := <-chReports:
-		if resp.Err != nil {
-			return nil, resp.Err
+
+	for _, trg := range e.workflow.Targets {
+		err := e.handleStep(ctx, ec, trg)
+		if err != nil {
+			e.logger.Errorf("error in handleTargets %v", err)
+			return
 		}
-		return resp.Value, nil
-	case <-ctx.Done():
-		return nil, ctx.Err()
 	}
 }
 
-func (e *Engine) handleTargets(ctx context.Context, resp values.Value) error {
-	e.logger.Debugw("handle targets")
-	inputs := map[string]values.Value{
-		"report": resp,
+func (e *Engine) handleStep(ctx context.Context, es *executionState, node Capability) error {
+	stepState := &stepState{
+		outputs: &stepOutput{},
 	}
+	es.steps[node.Ref] = stepState
 
-	var combinedErr error
-	for _, t := range e.targets {
-		e.logger.Debugw("sending to target", "target", t.typeStr, "inputs", inputs)
-		tr := capabilities.CapabilityRequest{
-			Inputs: &values.Map{Underlying: inputs},
-			Config: t.config,
-			Metadata: capabilities.RequestMetadata{
-				WorkflowID:          mockedWorkflowID,
-				WorkflowExecutionID: mockedExecutionID,
-			},
-		}
-		_, err := capabilities.ExecuteSync(ctx, t.capability, tr)
-		combinedErr = errors.Join(combinedErr, err)
+	// Let's get the capability. If we fail here, we'll bail out
+	// and try to handle it at the next execution.
+	cp, err := e.registry.Get(ctx, node.Type)
+	if err != nil {
+		return err
+	}
+
+	api, ok := cp.(capabilities.CallbackExecutable)
+	if !ok {
+		return fmt.Errorf("capability %s must be an action, consensus or target", node.Type)
 	}
-	return combinedErr
+
+	i, err := interpolateInputsFromState(node.Inputs, es)
+	if err != nil {
+		return err
+	}
+
+	inputs, err := values.NewMap(i.(map[string]any))
+	if err != nil {
+		return err
+	}
+
+	stepState.inputs = inputs
+
+	config, err := values.NewMap(node.Config)
```

Comment:
We could pre-compute some of these (config, workflow ID, getting from Registry) to avoid doing it on every execution.

Reply:
@bolekk Agree with config + workflow ID, but can we do this for getting from the registry? We can't guarantee that a remote capability will always be available for instance, hence why I thought it would make sense fetch it dynamically.

(Although, that raises the thought of: we actually need the capability to be the same since we call the RegisterToWorkflow hooks at the beginning of the workflow instantiation. Do you think this is a safe assumption to make?)

Reply:
As discussed offline - we will decide separately how to handle capability requirements.